components/patient-data/rest/src/main/java/org/phenotips/data/rest/internal/DefaultPatientsFetchResourceImpl.java

```diff
@@ -0,0 +1,222 @@
+@Singleton
+public class DefaultPatientsFetchResourceImpl extends XWikiResource implements PatientsFetchResource
+{
+    /** Jackson object mapper to facilitate array serialization. */
+    private static final ObjectMapper OBJECT_MAPPER = getCustomObjectMapper();
+
+    private static final String SERIALIZER_LABEL = "PrimaryEntitySerializer";
+
+    /** Logging helper object. */
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private QueryManager qm;
+
+    @Inject
+    private PatientRepository repository;
+
+    @Inject
+    private AuthorizationManager access;
+
+    @Inject
+    private UserManager users;
+
+    @Inject
+    private Container container;
+
+    @Override
+    public Response fetchPatients()
+    {
+        final Request request = this.container.getRequest();
+        // Get the internal and external IDs, if provided.
+        final List<Object> eids = request.getProperties("eid");
+        final List<Object> ids = request.getProperties("id");
+
+        this.logger.debug("Retrieving patient records with external IDs [{}] and patient IDs [{}] via REST", eids, ids);
+
+        // Build a set of patients from the provided external and/or internal ID data.
+        final ImmutableSet.Builder<PrimaryEntity> patientsBuilder = ImmutableSet.builder();
+
+        addEids(patientsBuilder, eids);
+        addIds(patientsBuilder, ids);
+
+        try {
+            // Generate JSON for all retrieved patients.
+            final String json = OBJECT_MAPPER.writeValueAsString(patientsBuilder.build());
+            return Response.ok(json, MediaType.APPLICATION_JSON_TYPE).build();
+        } catch (final JsonProcessingException ex) {
+            logger.warn("Failed to serialize patients [{}] to JSON: {}", eids, ex.getMessage());
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        }
+    }
+
+    /**
+     * Retrieves patient entities given a list of external patient IDs.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param eids a list of external patient IDs, as strings
+     */
+    private void addEids(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final List<Object> eids)
+    {
+        for (final Object eid : eids) {
+            if (StringUtils.isNotBlank((String) eid)) {
+                collectPatientsFromEid(patientsBuilder, eid);
+            }
+        }
+    }
+
+    /**
+     * Retrieves and collects patient entities that correspond to the provided external ID.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param eid an external patient ID, as string
+     */
+    private void collectPatientsFromEid(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final Object eid)
+    {
+        try {
+            final Query q = qm.createQuery("where doc.object(PhenoTips.PatientClass).external_id = :eid", Query.XWQL);
```

Comment:
Instead of running a query for every eid, you could just do:

```
q = qm.createQuery("from doc.object(PhenoTips.PatientClass) p where p.external_id in (:eids)", Query.XWQL);
q.bindValue("eids", eids);
q.execute()
```